### DISCUSSION TITLE
Can Ghidra propagate parameters definition to functions signature?

### DISCUSSION THREAD

**Discussion by karpiyon** [NationalSecurityAgency/ghidra#7198] [Q&A]
Hi,
After many year of an ameture IDA user I decided to check the magic of Ghidra.

I'm reading "The Ghidra Book The Definitive Guide" and in one of the examples it shows how to replace a function parameter type, in this case:

to replace param_3 to an addrinfo pointer in this definition:
`ssize_t get_page(undefined4 param_1,undefined4 param_2, int param_3)
`
(If you want the exact location it is on pg.172 in this sample "http_get_example_x86"
[](https://ghidrabook.com/static/GhidraBookCode.zip))

The reason is that if you follow the function definition and it's caller you find out that param_3 is really a pointer to an addrinfo struct.

It goes down like this:

```
undefined4 http_client(undefined4 param_1,undefined4 param_2)
{
  undefined4 uVar1;
  uVar1 = do_setup(param_1);
  uVar1 = get_page(param_1,param_2,uVar1);
  return uVar1;
}
```

and uVar1 (param_3 above) is defined in `do_setup` as `addrinfo` pointer like this:

```
addrinfo * do_setup(char *param_1)
{
  char *pcVar1;
  addrinfo local_34;
  addrinfo *local_14;
  int local_10;
  memset(&local_34,0,0x20);
  local_34.ai_family = 2;
  local_34.ai_socktype = 1;
  local_10 = getaddrinfo(param_1,"www",&local_34,&local_14);
  if (local_10 != 0) {
    pcVar1 = gai_strerror(local_10);
    fprintf(stderr,"getaddrinfo: %s\n",pcVar1);
                    /* WARNING: Subroutine does not return */
    exit(-1);
  }
  return local_14;
}
```

Now, going back to my question, it is clear now that param_3 in get_page definition is a pointer to addrinfo struct and I should replace it like this: 

`ssize_t get_page(undefined4 param_1,undefined4 param_2,int param_3)
`==>
`ssize_t get_page(undefined4 param_1,undefined4 param_2,addrinfo *param_3)
`

which makes the inner definition of any reference to param_3 a lot clearer.

```
  uVar1 = do_setup(param_1);
  uVar1 = get_page(param_1,param_2,uVar1);
```

uVar1 which is passed to get_page is a pointer to addrinfo as Ghidra deducted above therefore the definition/signature of get_page should reflect it.

So my basic question is this:
Why, or how can I make Ghidra automatically deduct that param_3 is actually a pointer to addrinfo in this case.
After all my tracing back the origin of param_3, in this case is very easy.
It is right there in Ghidra code. Why then, did it not propagate?
If a parameter is decompiled as a struct and later passed on to a function why is it that that function definition/signature does no reflect it?

Of course I can do this manually in this case but when you try to RE real code it is a lot harder and takes a lot more time.

Perhaps I am missing something or perhaps there is a plugin or an option I did not use

**Comment by ghidracadabra** [comment]
From the code browser, try Analysis->One Shot->Decompiler Parameter ID

**Comment by DualTachyon** [comment]
Meant to reply separately, not to a reply...

When you fix a signature, you should press "p" to commit the changes. This will trigger Ghidra to propagate some of the changes. It doesn't always do that but it works often
